WakaTime/WakaTime.swift

```diff
@@ -66,6 +68,86 @@ struct WakaTime: App {
         }
     }
     
+    private func getLatestVersion(completion: @escaping ((String?, Error?) -> Void)) {
+        struct Release: Decodable {
+            let tagName: String
+            private enum CodingKeys: String, CodingKey {
+                case tagName = "tag_name"
+            }
+        }
+        
+        let apiUrl = "https://api.github.com/repos/wakatime/wakatime-cli/releases/latest"
+        var request = URLRequest(url: URL(string: apiUrl)!, cachePolicy: .reloadIgnoringCacheData)
+        let lastModified = ConfigFile.getSetting(section: "internal", key: "cli_version_last_modified", internalConfig: true)
+        let currentVersion = ConfigFile.getSetting(section: "internal", key: "cli_version", internalConfig: true)
+        if let lastModified, currentVersion != nil {
+            request.setValue(lastModified, forHTTPHeaderField: "If-Modified-Since")
+        }
+        URLSession.shared.dataTask(with: request) { data, response, error in
+            guard let data, let httpResponse = response as? HTTPURLResponse else {
+                completion(nil, error)
+                return
+            }
+            if httpResponse.statusCode == 304 {
+                DispatchQueue.main.async {
+                    // Current version is still the latest version available
+                    completion(currentVersion, nil)
+                }
+            } else if let lastModified = httpResponse.value(forHTTPHeaderField: "Last-Modified"),
+                      let release = try? JSONDecoder().decode(Release.self, from: data) {
+                // Remote version successfully decoded
+                ConfigFile.setSetting(section: "internal", key: "cli_version_last_modified", val: lastModified, internalConfig: true)
+                ConfigFile.setSetting(section: "internal", key: "cli_version", val: release.tagName, internalConfig: true)
+                DispatchQueue.main.async {
+                    completion(release.tagName, nil)
+                }
+            } else {
+                DispatchQueue.main.async {
+                    // Unexpected response
```

Comment:
This might happen when the response looks ok, but the body is malformed (no data or whatever could happen). The caller is capable of handling the `(nil, nil)` case.

Reply:
We could use `async throws` and then only return one `nil` instead of two?

https://www.hackingwithswift.com/quick-start/concurrency/how-to-call-async-throwing-functions